### PR TITLE
LPS-154718 Fix style of client-extensions

### PR DESCRIPTION
**Pull request by edalgrin**
This PR fixes and improves the visual styles of the remote-apps (client-extensions) module.

ISSUE: https://issues.liferay.com/browse/LPS-154718
FIGMA: https://www.figma.com/file/VPkwjauHnA9jxBqUHPyc8j/lps-152023-extend-remote-apps?node-id=108%3A145593
STATUS:
localhost_8080_group_guest___control_panel_manage_p_p_id=com_liferay_client_extension_web_internal_portlet_ClientExtensionAdminPortlet p_p_lifecycle=0 p_p_state=maximized p_p_mode=view _com_liferay_client_extension_web_internal_portlet_Clie

### CODE REVIEW

**Comment by edalgrin** [comment]
ci:test:sf

**Comment by edalgrin** [comment]
Missing changes:

- Add a description for each type of client-extension
image

---

- Change the control-menu-title
image

---

- Replace the AlloyEditor used in the description field with CKEditor
image

---

- Add numbers in repeated field labels for custom-element type
image

---

- Modify view and create menu
image
image

---

- Generate language files (`ant build-lang` is not working for me)
image

**Comment by edalgrin** [comment]
One thing I noticed during this change is that `<liferay-ui:input-localized...` doesn't seem to have the `required` attribute. I managed to add a required flag to its wrapper `<aui:field-wrapper...` and it works with the HTML5 required field.

image

**Comment by edalgrin** [comment]
ci:test:relevant

**Comment by edalgrin** [comment]
ci:test:sf

**Comment by john-co** [comment]
[ Test fix ]
Portlet Category Name field changed to Widget Category Name - I'll push a test fix to this branch

**Exploratory Testing:**
[ UI inconsistencies ]
- The styling/format of error message for Name field is not the same as Lexicon/Clay for alert messages. And does not match figma doc.
Selection_318
- alert message is not adjacent and under the URL field
Selection_316
- After adding a Remote App, the alert message does not appear for invalid URL...I'm not sure the exact steps yet. Discovered this at the end of exploratory testing. 
image

**Comment by edalgrin** [comment]
Hey @john-co, thanks for taking a look at this!

> * The styling/format of error message for Name field is not the same as Lexicon/Clay for alert messages. And does not match figma doc.
>   Selection_318

That's because the current `<liferay-ui:input-localized...` doesn't seem to have the `required` attribute so I had to get _creative_ https://github.com/liferay-frontend/liferay-portal/pull/2594#issuecomment-1210475562

> * When URL is invalid and has atleast 1 character, alert message is not adjacent and under the URL field
>   Selection_316

I think it's possible to duplicate the error message and move them to the fields but it might require some changes from the "backend" side to apply the right conditionals

> * When URL field is blank, there is no error alert message
>   image

I'm not sure that the error-alert-message is even required by the design, we have it as default for this module, perhaps we could remove it if necessary

**Comment by edalgrin** [comment]
The biggest challenge we have in this change is the order of the fields. The design requires a certain logical order but the module has been built with performance as first objective, hence some common fields are grouped in a [single file](https://github.com/liferay/liferay-portal/blob/master/modules/apps/client-extension/client-extension-web/src/main/resources/META-INF/resources/admin/edit_client_extension_entry.jsp) while the others, specific for each type of client-extension, are distributed in [multiple files](https://github.com/liferay/liferay-portal/tree/master/modules/apps/client-extension/client-extension-web/src/main/resources/META-INF/resources/admin). This means we should either adapt the design to the development choice or find a new the way to adjust the module.

/cc @ugeortiz @izaera

**Comment by ugeortiz** [comment]
Thanks @edalgrin, if the order of the fields presents a problem, maybe we can think about a workaround. The main problem is the `properties`field, right?

**Comment by edalgrin** [comment]
ezgif com-gif-maker

**Comment by edalgrin** [comment]
Moved repeat-fields-buttons to avoid overlap with input-fields

image

**Comment by edalgrin** [comment]
Add type descriptions

image

File: modules/apps/client-extension/client-extension-web/src/main/java/com/liferay/client/extension/web/internal/display/context/util/CETLabelUtil.java [hunk 1]
```diff
@@ -32,6 +32,12 @@ public static String getAddLabel(
 		return _getLabel(httpServletRequest, "add-", type);
 	}
 
+	public static String getHelpNameLabel(
```
Comment (izaera) [inline modules/apps/client-extension/client-extension-web/src/main/java/com/liferay/client/extension/web/internal/display/context/util/CETLabelUtil.java:35]: This should probably be `getHelpLabel` (as `getNewLabel`, for example)

Comment (edalgrin) [reply modules/apps/client-extension/client-extension-web/src/main/java/com/liferay/client/extension/web/internal/display/context/util/CETLabelUtil.java:35]: I was following the example of `getTypeLabel` – is it correct to have the same name for the "function" and the "callback" ?

image

**Comment by john-co** [comment]
Hi @edalgrin , I remember pushing in a test fix to this branch but it may have been overwritten by last force push. If still applicable, please cherry pick this [test fix](https://github.com/john-co/liferay-portal/pull/482)

**Comment by edalgrin** [comment]
@john-co Oh I'm sorry I totally forgot about it, thank you for notice it!! I'll cherry-pick it right away

**Comment by edalgrin** [comment]
> The styling/format of error message for Name field is not the same as Lexicon/Clay for alert messages. And does not match figma doc.

So I tried to modify the tag `<liferay-ui:input-localized>` and add the `required` attribute to it to keep the consistency as required. The tag contains the AUI library so it could be relatively easy to add an `<aui:input required="true" .. >` inside it, and that's what I tried. But since the field has multiple inputs due to its multi-language characteristic it's hard to change it.

So I prefer to leave the field as it is at the moment and work on it separately in a new "add required attribute to input-localized" Story – perhaps someone else could work on it while I finish these changes?

**Comment by edalgrin** [comment]
Fix page title (it will be New Client Extension, once changed for the whole module)

image
image

**Comment by edalgrin** [comment]
One of the problems we are facing now is that the HTML required attribute (yellow one) doesn't open the closed Identity accordion when triggered.
ezgif com-gif-maker

To avoid that we want to remove the current HTML required attribute from the Name field (input:localized) and add a Java validation for it with the red alert


I tried to duplicate the method of the NewCustomElement interface and it works but that's not the right way, we already have the `name` declared in the CET.java file we just need to add the validation, but I'm not sure how to do it.

**Comment by john-co** [comment]
ci:test:bundle

**Comment by john-co** [comment]
ci:test:bundle

**Comment by john-co** [comment]
ci:test:relevant

**Comment by john-co** [comment]
Hi @edalgrin @ugeortiz , I noticed while reviewing automation test in https://github.com/liferay-frontend/liferay-portal/pull/2751 that we don't have Name as required field. Can we still expect Name field to be a required field? 

cc/ @bronsonsoda

**Comment by john-co** [comment]
Same thing for Custom Element. In addition, sometimes the error message is pointing out the wrong field. ie: I corrected the URL fields but the error message still says something is wrong with the URL fields. 

https://user-images.githubusercontent.com/35235266/192396642-41f23eea-3da5-40c9-a7eb-d2ee2322c575.mp4

**Comment by john-co** [comment]
ci:test:bundle

**Comment by edalgrin** [comment]
ci:test:sf

**Comment by edalgrin** [comment]
ci:test:relevant

**Comment by edalgrin** [comment]
ci:test:relevant

**Comment by john-co** [comment]
ci:test:frontend-remote-apps

**Comment by edalgrin** [comment]
Fixed conflicts, retrying tests

**Comment by edalgrin** [comment]
ci:test:relevant

**Comment by john-co** [comment]
ci:test:relevant

**Comment by john-co** [comment]
ci:test:frontend-remote-apps

**Comment by dsanz** [comment]
ci:reevaluate:1327401_6706

**Comment by dsanz** [comment]
Hi @john-co @mateusralv could you please help me to understand which tests related to ci:test:frontend-remote-apps are failing?

We need to merge these changes as this is blocking further developments related to client extension creation

**Comment by mateusralv** [comment]
I'm taking a look 🕵🏽

**Comment by mateusralv** [comment]
Hey @edalgrin and @dsanz , most of the faults are not related. ([Comparison run](https://testray.liferay.com/home/-/testray/runs/compare?formDate=1667937185128&testrayRunIdA=2033394646&testrayRunIdB=2030602210))

I found only one failure related: 
1. LocalFile.RemoteAppsRepeatableFields#JSURLCanBeRepeated
PRreview

Added two spaces in the element with 'JS URL'. 
Before it was 'JS URL' after the PR: ' JS URL '. With that, the test is breaking. I can fix the test by adding the spaces in the path in a commit to your PR if you give me permission to do so 😄 .

Failures not unique:
1. LocalFile.AssetsCustomAsset#CanCreateMultipleCustomAssetReport
3. LocalFile.BlogsRemoteStaging#CanPublishUpdatedCoverImage
4. LocalFile.CalendarConfiguration#ViewCalendarTimeZoneDefaults
5. LocalFile.CollectionFilter#FilterResultsOfCollectionPageBasedOnDynamicCollectionByContainsAllCategories
6. LocalFile.ObjectPermission#CanEditItsOwnEntryWithOnlyAddPermission
7. LocalFile.ObjectPortlet#CannotViewOtherUsersEntryWithOnlyAddPermission
8. LocalFile.Ratings#CanRevokeStarsFromKB
9. LocalFile.Staging#StagingMultipleDM
10. LocalFile.AutoTranslationGoogle#CanTranslateWCToJapaneseWithGoogle

**Comment by john-co** [comment]
Thanks @mateusralv ! yes we can add a test fix for this.

Unfortunately there are other related test failures from test suite [frontend-remote-apps](https://github.com/liferay-frontend/liferay-portal/pull/2594#issuecomment-1301426187)


**[Related]**
- RemoteApps#HasMandatoryFieldErrorMessagesWhenEmpty - Test Fix update the error text assertion
- [LocalFile.RemoteApps#ErrorAppearsWhenHTMLBlank](https://test-1-11.liferay.com/job/test-portal-acceptance-pullrequest-downstream(master)/122377//testReport/com.liferay.poshi.runner/PoshiRunner/test_LocalFile_RemoteApps_ErrorAppearsWhenHTMLBlank_) - [Poshi Report](https://testray.liferay.com/reports/production/logs/2022-11/test-1-32/test-portal-acceptance-pullrequest(master)/6707/functional-tomcat90-mysql57-jdk8/0/4/LocalFile.RemoteApps_ErrorAppearsWhenHTMLBlank/index.html.gz) - [Poshi Summary](https://testray.liferay.com/reports/production/logs/2022-11/test-1-32/test-portal-acceptance-pullrequest(master)/6707/functional-tomcat90-mysql57-jdk8/0/4/LocalFile.RemoteApps_ErrorAppearsWhenHTMLBlank/summary.html.gz) - Test Fix update the error text assertion
- [LocalFile.RemoteApps#HTMLElementConatinsInvlaidCharacters](https://test-1-31.liferay.com/job/test-portal-acceptance-pullrequest-downstream(master)/293086//testReport/com.liferay.poshi.runner/PoshiRunner/test_LocalFile_RemoteApps_HTMLElementConatinsInvlaidCharacters_) - [Poshi Report](https://testray.liferay.com/reports/production/logs/2022-11/test-1-32/test-portal-acceptance-pullrequest(master)/6707/functional-tomcat90-mysql57-jdk8/0/5/LocalFile.RemoteApps_HTMLElementConatinsInvlaidCharacters/index.html.gz) - [Poshi Summary](https://testray.liferay.com/reports/production/logs/2022-11/test-1-32/test-portal-acceptance-pullrequest(master)/6707/functional-tomcat90-mysql57-jdk8/0/5/LocalFile.RemoteApps_HTMLElementConatinsInvlaidCharacters/summary.html.gz) - Impediment Bug, expected error `html element has an invalid character` (white space), but no error is thrown.
- [LocalFile.RemoteApps#HTMLElementMustHaveOneHyphen](https://test-1-31.liferay.com/job/test-portal-acceptance-pullrequest-downstream(master)/293086//testReport/com.liferay.poshi.runner/PoshiRunner/test_LocalFile_RemoteApps_HTMLElementMustHaveOneHyphen_) - [Poshi Report](https://testray.liferay.com/reports/production/logs/2022-11/test-1-32/test-portal-acceptance-pullrequest(master)/6707/functional-tomcat90-mysql57-jdk8/0/5/LocalFile.RemoteApps_HTMLElementMustHaveOneHyphen/index.html.gz) - [Poshi Summary](https://testray.liferay.com/reports/production/logs/2022-11/test-1-32/test-portal-acceptance-pullrequest(master)/6707/functional-tomcat90-mysql57-jdk8/0/5/LocalFile.RemoteApps_HTMLElementMustHaveOneHyphen/summary.html.gz) - Impediment bug, error message is not thrown `HTML element name must contain at least one hyphen`
- [LocalFile.RemoteApps#HTMLElementMustStartWithLowercase](https://test-1-31.liferay.com/job/test-portal-acceptance-pullrequest-downstream(master)/293086//testReport/com.liferay.poshi.runner/PoshiRunner/test_LocalFile_RemoteApps_HTMLElementMustStartWithLowercase_) - [Poshi Report](https://testray.liferay.com/reports/production/logs/2022-11/test-1-32/test-portal-acceptance-pullrequest(master)/6707/functional-tomcat90-mysql57-jdk8/0/5/LocalFile.RemoteApps_HTMLElementMustStartWithLowercase/index.html.gz) - [Poshi Summary](https://testray.liferay.com/reports/production/logs/2022-11/test-1-32/test-portal-acceptance-pullrequest(master)/6707/functional-tomcat90-mysql57-jdk8/0/5/LocalFile.RemoteApps_HTMLElementMustStartWithLowercase/summary.html.gz) - Impediment Bug, error message not thrown `HTML element name must start with a lowercase letter`

Reference: [FIGMA file](https://www.figma.com/file/VPkwjauHnA9jxBqUHPyc8j/lps-152023-extend-remote-apps?node-id=61%3A130305) - receive error information

**Comment by john-co** [comment]
cc/ @mateusralv, the test fix for RemoteAppsRepeatableFields#JSURLCanBeRepeated looks like we just need to update the key_fieldLabelName to `JS URL`. Before the field label name was just `URL`. I manually tested this and it works.

**Comment by mateusralv** [comment]
Yes @john-co, I made the modifications and ran the poshi test in Bundle PR and it worked, the repair for the other tests, do you think better to send this PR or can I create a ticket and send a new PR?

**Comment by edalgrin** [comment]
I was taking a look at the reported impediment bugs but I don't understand what's the error with the tests.

image
image
image

**Comment by john-co** [comment]
Thanks @edalgrin , from your screenshots it does look like the error messages are correct so QA can look into including a test fix. @mateusralv will provide an update on it.

However, it looks like the HTML Element Name field was highlighted in green vs. the expected red like in JS URL. Will this be fixed?

image

Also, wanted to confirm what we're doing here since things don't exactly match the figma file. It looks like we're not going to have the error message for HTML Element Name directly under the incorrect field? It will only be on top of the page? Are there plans to have the error message be directly under the field similar to the figma design? You could see here in previous comment that it looked like we had some error indicator under the HTML field
https://github.com/liferay-frontend/liferay-portal/pull/2594#issuecomment-1258751283

**Comment by john-co** [comment]
Also would like feedback for this question, cc/ @edalgrin @ugeortiz 

https://github.com/liferay-frontend/liferay-portal/pull/2594#issuecomment-1258713300

**Comment by john-co** [comment]
ci:test:frontend-remote-apps

**Comment by mateusralv** [comment]
ci:test:frontend-remote-apps

**Comment by john-co** [comment]
ci:test:frontend-remote-apps

**Comment by john-co** [comment]
[screencapture-localhost-8080-group-guest-control-panel-manage-2022-11-15-15_32_38.pdf](https://github.com/liferay-frontend/liferay-portal/files/10016873/screencapture-localhost-8080-group-guest-control-panel-manage-2022-11-15-15_32_38.pdf)

I couldn't reproduce the messages from https://github.com/liferay-frontend/liferay-portal/pull/2594#issuecomment-1310538505

**Comment by mateusralv** [comment]
Hi @edalgrin , I was able to reproduce these pages in Bundle PR and created a fix for a test that is failing, but I found 2 failures that seem to be unique: 

1. [LocalFile.RemoteApps#HTMLElementMustStartWithLowercase](https://testray.liferay.com/home/-/testray/case_results/2054520977?redirect=https%3A%2F%2Ftestray.liferay.com%2Fhome%2F-%2Ftestray%2Fcase_results%2Findex%3FtestrayBuildId%3D2054520579%26testrayRunId%3D2054520630%26cur%3D1%26delta%3D200%26orderByCol%3Dstatus_sortable%26orderByType%3Dasc)

Steps to reproduce: 
Add a Add Custom Element > write "Vanilla"  on `HTML Element Name` input.
on the right: master | on the left: bundle`s PR



https://user-images.githubusercontent.com/44584050/202561540-3247003c-855d-4b2a-a512-4413b31ba3de.mp4

same as:
2. [LocalFile.RemoteApps#HTMLElementMustHaveOneHyphen](https://testray.liferay.com/home/-/testray/case_results/2054520959?redirect=https%3A%2F%2Ftestray.liferay.com%2Fhome%2F-%2Ftestray%2Fcase_results%2Findex%3FtestrayBuildId%3D2054520579%26testrayRunId%3D2054520630%26cur%3D1%26delta%3D200%26orderByCol%3Dstatus_sortable%26orderByType%3Dasc)
Steps to reproduce: 
Add a Add Custom Element > write "vanilla.counter"  on `HTML Element Name` input.


3. [LocalFile.RemoteApps#HTMLElementConatinsInvlaidCharacters](https://testray.liferay.com/home/-/testray/case_results/2054520941?redirect=https%3A%2F%2Ftestray.liferay.com%2Fhome%2F-%2Ftestray%2Fcase_results%2Findex%3FtestrayBuildId%3D2054520579%26testrayRunId%3D2054520630%26cur%3D1%26delta%3D200%26orderByCol%3Dstatus_sortable%26orderByType%3Dasc)
Steps to reproduce: 
Add a Add Custom Element > write "vanilla counter"  on `HTML Element Name` input.

4. [LocalFile.RemoteApps#ErrorAppearsWhenHTMLBlank](https://testray.liferay.com/home/-/testray/case_results/2054520923?redirect=https%3A%2F%2Ftestray.liferay.com%2Fhome%2F-%2Ftestray%2Fcase_results%2Findex%3FtestrayBuildId%3D2054520579%26testrayRunId%3D2054520630%26cur%3D1%26delta%3D200%26orderByCol%3Dstatus_sortable%26orderByType%3Dasc) Test fix i'll be created.


cc: @john-co, @dsanz

**Comment by edalgrin** [comment]
If the Master is actually on the left and the PR on the right, it's correct!

At the moment we have 2 validations process in the page:
- The first one (JS) happens either when you
  - focus and _blur_ an empty required field
  - submit the page/form without filling all the required fields (except `Name`)
- The second one (SERVER) happens after you submit the page/form and it checks that:
  - The `Name` field is not empty
  - The `HTML Element Name` field is not Uppercase
  - The `HTML Element Name` field does not contain any symbol but `-`
  - The `JS Url` field contains a valid url

In any case, I know everything is very connected but these are errors for Client Extensions **Custom Element** type, in this PR/phase we should check mostly for the IFrame type.

**Comment by john-co** [comment]
ci:test:frontend

**Comment by john-co** [comment]
ci:test:frontend-remote-apps

**Comment by dsanz** [comment]
ci:reevaluate:1397401_1778

**Comment by mateusralv** [comment]
ci:test:frontend-remote-apps

**Comment by john-co** [comment]
ci:test:frontend-remote-apps

**Comment by mateusralv** [comment]
New tests failed from[ these results](https://github.com/liferay-frontend/liferay-portal/pull/2594#issuecomment-1322530323). I’m creating the fix to update the PR.

**Comment by mateusralv** [comment]
ci:test:frontend-remote-apps

**Comment by mateusralv** [comment]
ci:test:relevant

**Comment by mateusralv** [comment]
I'm taking a look 👓

**Comment by mateusralv** [comment]
Hi @edalgrin the failures are not related to your PR. 😄 
[comparison run ](https://testray.liferay.com/home/-/testray/runs/compare?formDate=1669207611987&testrayRunIdA=2071633178&testrayRunIdB=2071566264)

**Comment by mateusralv** [comment]
The failures are not unique, but we can run ci:test:relevant after pull, rebase and conflict fix, for double check.

**Comment by edalgrin** [comment]
ci:test:relevant

**Comment by mateusralv** [comment]
ci:test:relevant

**Comment by mateusralv** [comment]
ci:test:relevant

**Comment by mateusralv** [comment]
ci:test:relevant

**Comment by john-co** [comment]
ci:test:frontend-remote-apps:norebase

**Comment by john-co** [comment]
ci:test:relevant:norebase

**Comment by john-co** [comment]
Rebase attempt and testing on https://github.com/john-co/liferay-portal/pull/490

**Comment by john-co** [comment]
I'm still running into a compile error after attempted rebase:

```
  [exec] > Task :apps:client-extension:client-extension-service:compileJava FAILED
     [exec] 
     [exec] > Task :apps:client-extension:client-extension-uad:jar
     [exec] > Task :apps:client-extension:client-extension-uad:autoUpdateXml SKIPPED
     [exec] /opt/dev/projects/github/liferay-portal/modules/apps/client-extension/client-extension-service/src/main/java/com/liferay/client/extension/service/impl/ClientExtensionEntryLocalServiceImpl.java:484: error: incompatible types: long cannot be converted to java.lang.String
     [exec]           companyId, externalReferenceCode);
 ```
 
 @edalgrin would you be able to take a look?

**Comment by edalgrin** [comment]
I'm going to take a look, thanks!

**Comment by edalgrin** [comment]
Sorry I was on sick leave so I could take a deeper look but from what I've seen it seemed a JAVA issue so I spoke with @dsanz and he might have – any news?

**Comment by dsanz** [comment]
See https://github.com/liferay-frontend/liferay-portal/pull/2894 followup PR

**Pull request closed without merging**